test/decl/protocol/protocols_with_self_or_assoc_reqs.swift

```diff
@@ -0,0 +1,128 @@
+// RUN: %target-typecheck-verify-swift
+
+//===----------------------------------------------------------------------===//
+// Use of protocols with Self or associated type requirements
+//===----------------------------------------------------------------------===//
+
+protocol P1 {
+  associatedtype Q
+  func returnSelf() -> Self
+  func returnAssoc() -> Q
+}
+
+struct S1: P1 {
+  typealias Q = Int
+  func returnSelf() -> Self { self }
+  func returnAssoc() -> Q { 0 }
+}
+
+let p1: P1 = S1()
+_ = p1.returnSelf() // ok
+_ = p1.returnAssoc() // expected-error {{member 'returnAssoc' cannot be used on value of protocol type 'P1'; use a generic constraint instead}}
```

Comment:
It would be really great if we could change this to something more detailed like `member 'returnAssoc' cannot be used on value of protocol type 'P1'; because it returns associated type 'Q', it may only be used on a concrete conforming type`, since if the proposal is accepted users will encounter it _much_ more often

Reply:
On that note, it might be better to update the related educational note to explain why these members can't be used, rather than remove the educational note altogether.

Reply:
Turns out there's already at least one use of `Self` that's allowed, which is difficult to explain. I've left #33736 as a draft so that the educational note can evolve in parallel with these changes.